services/subscriber/service.go

```diff
@@ -494,3 +603,78 @@ func (b *balancewriter) Statistics(tags map[string]string) []models.Statistic {
 	}
 	return statistics
 }
+
+type dbrp struct {
+	db, rp string
+}
+
+type subscriberMap struct {
+	// m is a two-level map to make finding the db,rp match efficient
+	m map[dbrp]map[string]*chanWriter
+}
+
+// upsert upserts a channel writer for subEntry
+func (s *subscriberMap) upsert(se subEntry, cw *chanWriter) {
+	if s.m == nil {
+		s.m = make(map[dbrp]map[string]*chanWriter)
+	}
+	key1 := dbrp{
+		db: se.db,
+		rp: se.rp,
+	}
+	if s.m[key1] == nil {
+		s.m[key1] = make(map[string]*chanWriter)
+	}
+	s.m[key1][se.name] = cw
+}
+
+func (s *subscriberMap) forEachChanWriter(f func(cw *chanWriter)) {
+	for _, m1 := range s.m {
+		for _, cw := range m1 {
+			f(cw)
+		}
+	}
+}
+
+// forEachDbRp runs f for all the channel writers with subscriptions matching a given write request
+func (s *subscriberMap) forEachDbRp(db, rp string, f func(cw *chanWriter)) {
+	key1 := dbrp{
+		db: db,
+		rp: rp,
+	}
+	for _, cw := range s.m[key1] {
```

Comment:
All this subscriberMap code is so I can do a map lookup for the subscriptions that matter, rather than iterating through all the subscriptions. Thoughts on whether this is a premature optimization?